internal/apps/dop/providers/rule/actions/pipeline/provider.go

```diff
@@ -0,0 +1,203 @@
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/bundle"
+	"github.com/erda-project/erda/internal/apps/dop/providers/projectpipeline"
+	"github.com/erda-project/erda/internal/apps/dop/services/apierrors"
+	"github.com/erda-project/erda/internal/apps/dop/services/branchrule"
+	"github.com/erda-project/erda/internal/apps/dop/services/pipeline"
+	"github.com/erda-project/erda/internal/pkg/diceworkspace"
+	"github.com/erda-project/erda/pkg/common/apis"
+)
+
+type provider struct {
+	bdl                *bundle.Bundle
+	ProjectPipelineSvc *projectpipeline.ProjectPipelineService `autowired:"erda.dop.projectpipeline.ProjectPipelineService"`
+	PipelineSource     sourcepb.SourceServiceServer            `autowired:"erda.core.pipeline.source.SourceService" required:"true"`
+	PipelineDefinition definitionpb.DefinitionServiceServer    `autowired:"erda.core.pipeline.definition.DefinitionService" required:"true"`
+	pipeline           *pipeline.Pipeline
+	branchRule         *branchrule.BranchRule
+}
+
+func (p *provider) WithPipelineSvc(svc *pipeline.Pipeline) {
+	p.pipeline = svc
+}
+
+func (p *provider) WithBranchRule(branchRule *branchrule.BranchRule) {
+	p.branchRule = branchRule
+}
+
+type Interface interface {
+	CreatePipeline(env map[string]interface{}) (string, error)
+	WithPipelineSvc(svc *pipeline.Pipeline)
+	WithBranchRule(branchRule *branchrule.BranchRule)
+}
+
+func (p *provider) Init(ctx servicehub.Context) error {
+	p.bdl = bundle.New(bundle.WithPipeline())
+	return nil
+}
+
+type PipelineConfig struct {
+	PipelineYml    string
+	PipelineYmlStr string
+	AppID          uint64
+	RefName        string
+	UserID         string
+	Path           string
+	FileName       string
+}
+
+func (p *provider) CreatePipeline(env map[string]interface{}) (string, error) {
+	gitPush, ok := env["git_push"]
+	if !ok {
+		return "", fmt.Errorf("empty git_push config")
+	}
+	v, ok := gitPush.(map[string]interface{})["pipelineConfig"]
+	if !ok {
+		return "", fmt.Errorf("empty pipeline config")
+	}
+	// parse pipeline config
+	b, err := json.Marshal(v)
+	if err != nil {
+		return "", err
+	}
+
+	var req PipelineConfig
+	if err := json.Unmarshal(b, &req); err != nil {
+		return "", err
+	}
+
+	app, err := p.bdl.GetApp(req.AppID)
+	if err != nil {
+		return "", err
+	}
+
+	// create pipeline
+	ctx := context.Background()
+	definitionID, err := p.GetPipelineDefinitionID(apis.WithUserIDContext(ctx, req.UserID), app, req.RefName, req.Path, req.FileName, req.PipelineYmlStr)
+	if err != nil {
+		logrus.Errorf("failed to bind definition %v", err)
+	}
+
+	reqPipeline := &apistructs.PipelineCreateRequest{
+		AppID:              req.AppID,
+		Branch:             req.RefName,
+		Source:             apistructs.PipelineSourceDice,
+		PipelineYmlSource:  apistructs.PipelineYmlSourceGittar,
+		PipelineYmlContent: req.PipelineYmlStr,
+		AutoRun:            true,
+		UserID:             req.UserID,
+	}
+	v2, err := p.pipeline.ConvertPipelineToV2(reqPipeline)
```

Comment:
I think these create logic is very similar to app pipeline. maybe code should be reused.

Reply:
they are in endpoints, any idea how we can avoid circular dependence.